conda_index/index/__init__.py

```diff
@@ -636,17 +640,7 @@ def index(
         logging_context = utils.LoggingContext(level, loggers=[__name__])
 
         with logging_context:
-            if not self._subdirs:
-                detected_subdirs = {
-                    subdir.name
-                    for subdir in os.scandir(self.channel_root)
-                    if subdir.name in utils.DEFAULT_SUBDIRS and subdir.is_dir()
-                }
-                log.debug("found subdirs %s" % detected_subdirs)
-                self.subdirs = subdirs = sorted(detected_subdirs | {"noarch"})
-            else:
-                self.subdirs = subdirs = sorted(set(self._subdirs))
-                log.warn("Indexing %s does not include 'noarch'", subdirs)
+            subdirs = self.detect_subdirs()
```

Comment:
This would be cleaner if it were `self._subdirs = _detect_subdirs(self.channel_root)`.  Then the `_detect_subdirs` function wouldn't have to have any side effects (like updating `self._subdirs` or `self.subdirs`).

It would be less confusing to not have both `self._subdirs` and `self.subdirs`.  To my eyes, I can't see a reason why we don't stick with just `self._subdirs`.

It's odd to me to update a param like this, esp since it's only used in one other place in this function to calculate the len.

```suggestion
            if not self._subdirs:
                self._subdirs = _detect_subdirs(self.channel_root)
```

If you accept the suggestion, you'll also have to update the one place where `self.subdirs` is referenced later.

```python
                    def extract_args():
                        for subdir in self.subdirs:
```

Reply:
`_subdirs` is assigned in the `ChannelIndex()` constructor if you pass the `-s` argument. `self.subdirs` should definitely be removed.